iceoryx2-cal/src/shm_allocator/mod.rs

```diff
@@ -13,44 +13,147 @@
 pub mod bump_allocator;
 pub mod pool_allocator;
 
-use std::{alloc::Layout, ptr::NonNull};
+use std::{alloc::Layout, fmt::Debug, ptr::NonNull};
 
 pub use iceoryx2_bb_elementary::allocator::AllocationError;
 use iceoryx2_bb_elementary::{allocator::BaseAllocator, enum_gen};
 
-pub trait ShmAllocatorConfig: Copy + Default {}
+/// Trait that identifies a configuration of a [`ShmAllocator`].
+pub trait ShmAllocatorConfig: Copy + Default + Debug {}
 
-#[derive(Clone, Copy, Eq, PartialEq, Debug)]
-pub struct PointerOffset(usize);
+/// Defines the [`SegmentId`] of a [`SharedMemory`](crate::shared_memory::SharedMemory)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct SegmentId(u8);
 
-impl PointerOffset {
-    pub fn new(value: usize) -> PointerOffset {
+impl SegmentId {
+    /// Creates a new [`SegmentId`] from a given value.
+    pub fn new(value: u8) -> Self {
         Self(value)
     }
 
-    pub fn value(&self) -> usize {
+    /// Returns the underlying value of the [`SegmentId`]
+    pub fn value(&self) -> u8 {
         self.0
     }
+
+    /// Returns the maximum value the [`SegmentId`] supports.
+    pub const fn max_segment_id() -> u8 {
+        u8::MAX
+    }
 }
 
-enum_gen! { ShmAllocationError
+const NUMBER_OF_BITS_IN_BYTE: usize = 8;
+
+/// An offset to a [`SharedMemory`](crate::shared_memory::SharedMemory) address. It requires the
+/// [`SharedMemory::payload_start_address()`](crate::shared_memory::SharedMemory::payload_start_address())
+/// of the corresponding [`SharedMemory`](crate::shared_memory::SharedMemory) to be converted into
+/// an actual pointer.
+///
+/// Contains the offset and the corresponding [`SegmentId`].
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub struct PointerOffset(u64);
+
+impl PointerOffset {
+    /// Creates a new [`PointerOffset`] from the given offset value with the [`SegmentId`] == 0.
+    pub fn new(offset: usize) -> PointerOffset {
+        const SEGMENT_ID: u64 = 0;
+        Self(
+            (offset as u64) << (core::mem::size_of::<SegmentId>() * NUMBER_OF_BITS_IN_BYTE)
+                | SEGMENT_ID,
+        )
+    }
+
+    /// Sets the [`SegmentId`] of the [`PointerOffset`].
+    pub fn set_segment_id(&mut self, value: SegmentId) {
+        self.0 |= value.0 as u64;
```

Comment:
```suggestion
        self.0 &= !((1u64 << SegmentIdUnderlyingType::BITS) - 1);
        self.0 |= value.0 as u64;
```

Reply:
I do not understand what kind of problem this solves.

Reply:
Because the bits need to be cleared before they are set, else this could happen
```rs
foo.set_segment_id(7); // results in 7
foo.set_segment_id(8); // results in 7 | 8 -> 15
```

It's a public function and can therefore be called multiple times.

Reply:
Oh wow ... you are right. I moved the PointerOffset into a separate file and created also unit tests verifying that this does work now.